wayland-server/src/dispatch.rs

```diff
@@ -252,11 +259,15 @@ impl<I: Resource + 'static, U: Send + Sync + 'static, D: Dispatch<I, U> + 'stati
 
     fn destroyed(
         &self,
+        handle: &wayland_backend::server::Handle,
         data: &mut D,
         client_id: wayland_backend::server::ClientId,
         object_id: wayland_backend::server::ObjectId,
     ) {
-        <D as Dispatch<I, U>>::destroyed(data, client_id, object_id, &self.udata)
+        let dhandle = DisplayHandle::from(handle.clone());
+        let resource = I::from_id(&dhandle, object_id).unwrap();
```

Comment:
Given we are invoking it on a dead object, this will produce and inert handle, from which the user data cannot be accessed. Is this what we want here?

Reply:
Probably not that specific behavior.

I'd however argue that wayland-backend changing it's behavior around object destruction to allow the user data to be accessible would be useful.

At the moment we have a few handlers in smithay which require tracking every protocol object that has been seen on the past. https://github.com/Smithay/smithay/blob/master/src/wayland/shell/xdg/handlers/surface/toplevel.rs#L125 since object IDs are still equal even after destruction, this can work to recover the object.

However even if the workaround suffices, this type of pattern is going to be needed for every protocol object type if we want to consider eliminating IsAlive in smithay or heavily reducing it's presence.

Reply:
To do that properly, we'd need to change the API of the `ObjectData` trait so that `destroyed()` takes an `Arc<Self>` instead of a `&self`, and use the doc-hidden `__set_object_data` method of the newly created resource to insert the `self` in it.

That'll be breaking change for both wayland-backend and wayland-server, though I'm thinking if we're doing that server side, we should also do it client side.

Reply:
I've started working on that. I do agree the client should have the same changes but I will do the client in a later pull request.